xdsl/dialects/riscv_scf.py

```diff
@@ -189,6 +186,32 @@ def parse(cls, parser: Parser) -> Self:
         return cls(lb, ub, step, iter_arg_operands, body)
 
 
+@irdl_op_definition
+class ForOp(ForRofBaseOp):
+    """
+    A for loop, counting up from lb to ub by step each iteration.
+    """
+    name = "riscv_scf.for"
+
+
+@irdl_op_definition
+class RofOp(ForRofBaseOp):
+    """
+    Reverse Order For loop.
+
+    MLIRs for loops have the implicit constraint that lb < ub, and step > 0.
+
+    In order to express loops that count down in higher-level IR, the rof op
+    is needed.
+
+    ROF has the semantics of going from lb to ub, decrementing by step each time.
+
+    In order to convert a for to a rof, one needs to switch lb and ub.
+    (for the normalized case that (ub - lb) % step == 0)
```

Comment:
Just noticed that we do not do any sanity checking for the bounds and step, or at least not as extensive as in `scf.for` and definitely lacking overall against MLIR.

Just noting this for now, I'm OK to go as we are!

Reply:
Not sure how well that works, as this may not be compile time information?

Reply:
It doesn't matter much ATM, there are other inconsistencies between the various flavours of our `for`s, but this is a good start; that's where I'm getting at.